First pass on proxy peering docs (#15790)
gravitational/teleport: docs/pages/management/operations/proxy-peering.mdx

---
title: Proxy Peering Migration
description: How to upgrade an existing Teleport cluster to Proxy Peering mode.
---

## Prerequisites

An existing Teleport Enterprise cluster. See our [getting started guide](../../deploy-a-cluster/teleport-enterprise/getting-started.mdx) for help
setting up a Teleport cluster.

All components in the cluster should be running Teleport `10.0` or later. See our
[upgrade procedure](./upgrading.mdx) for help upgrading a Teleport cluster.

Teleport Proxy instances must be able to reach each other over the network on port
3021 by default. Ensure there are no firewall policies that would block communication
between instances.

## Step 1/3. Enable Proxy Peering
Update your cluster's Auth Service configuration to set the tunnel strategy type
to `proxy_peering`.

```yaml
auth_service:
    tunnel_strategy:
        type: proxy_peering
        agent_connection_count: 1
```

This setting will indicate to agents that they are only required to connect to 1
Teleport Proxy instance as specified by the `agent_connection_count` field.

## Step 2/3. Restart the Auth Services

> Restart all Teleport Auth Services running in the cluster to apply the new auth service

Reviewer: ```suggestion
Restart all Teleport Auth Services running in the cluster to apply the new Auth Service
```

configuration.

## Step 3/3. Restart the Proxy Services
Restart all Teleport Proxy instances running in the cluster in order for Proxy Peering

> specific services to start up.

Reviewer: We should either say "Proxy Peering-specific" or rephrase so we won't need a compound modifier, e.g., "in order to start services related to Proxy Peering."

---
title: Proxy Peering Migration
description: How to upgrade an existing Teleport cluster to proxy peering mode.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="10.0"
>
  TLS routing is available starting from Teleport `10.0`.
</Details>

> ## Requirements

Reviewer: Let's rename this "Prerequisites" to be more consistent with other docs guides.

> ## Requirements

Reviewer: Are there any other requirements? E.g., is this guide for an organization with a pre-existing self-hosted Teleport cluster?

Should the cluster already be configured to use multiple Proxy Service instances?

Author: I will update the requirements for having an existing teleport cluster.

As for multiple proxy service instances it is not a requirement, however there wouldn't be a reason to enable proxy peering with a single proxy instance. So what do you think about some guidance on the benifits/use cases for proxy peering above this section?

> Teleport Proxy servers must be able to reach each other over the network. Ensure

Reviewer: ```suggestion
Teleport Proxy Service instances must be able to reach each other over the network. Ensure
```

Author: Replaced with just `instances` rather than `Service instances` for consistency.

> there are no firewall policies that would block proxy to proxy communication.

Reviewer: ```suggestion
there are no firewall policies that would block communication between instances.
```

> there are no firewall policies that would block proxy to proxy communication.

Reviewer: How would a reader know whether there are firewall policies that would block communication between Proxy Service instances?

Author: What to look for specifically? Some policy that would not allow inbound connections on port 3021 (default port for proxy peering). When Proxy Peering is not enabled there is no communication between Proxy instances so someone with an existing cluster may be blocking this traffic by default.

Author: Specified port here [2224ced](https://github.com/gravitational/teleport/pull/15790/commits/2224ced7bb3279a5786cb51463b943710580bd03)

## Step 1/4. Upgrade to teleport `10.0`

> Download Teleport `10.0` or later from the [downloads page](https://goteleport.com/teleport/download) or your enterprise portal

Reviewer: Since these instructions are environment specific, I think we should remove this section and include "Teleport 10.0" in the Prerequisites section. Otherwise, there are a lot of use cases we don't cover in this guide, e.g., Kubernetes deployments. Also, which Teleport services need to be on 10.0?

Author: Moved to prereqs here [593d6ac](https://github.com/gravitational/teleport/pull/15790/commits/593d6aca044d9dcfe96fd28099dbcedc9e2d431e)

and follow the standard [upgrade procedure](https://goteleport.com/docs/setup/operations/upgrading/).

> ## Step 2/4. Enable proxy peering

Reviewer: Capitalization

> Update your cluster's auth server configuration to set the tunnel strategy type to `proxy_peering`

Reviewer: ```suggestion
Update your cluster's Auth Service configuration to set the tunnel strategy type to `proxy_peering`
```

```yaml
auth_service:
    tunnel_strategy:
        type: proxy_peering
        agent_connection_count: 1
```

> This setting will indicate to agents that they are only required to connect to 1 proxy server as specified by the

Reviewer: same as other comments re: "proxy server" and "agents"

`agent_connection_count` field.

> ## Step 3/4. Restart auth services

Reviewer: ```suggestion
## Step 3/4. Restart the Auth Service
```

> Restart all auth services running in the cluster to apply the new auth service configuration

Reviewer: What's the best way to do this? If you have multiple Auth Service instances, can you restart them in any order?

Author: What do you mean by best way to do this? My initial reaction is it depends how teleport is deployed. If there are multiple auth services the order doesn't matter as long as the configs are the same.

## Step 4/4. Restart proxy services

> Restart all proxy services running in the cluster in order for proxy peering specific services to start up.

Reviewer: What's the best way to do this? Does order matter?

Author: Only order that matters here is that auth services are done before proxy services.

---
title: Proxy Peering Migration
description: How to upgrade an existing Teleport cluster to Proxy Peering mode.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="10.0"
>

>   TLS routing is available starting from Teleport `10.0`.

Reviewer 2: TLS routing?

---
title: Proxy Peering Migration
description: How to upgrade an existing Teleport cluster to Proxy Peering mode.
---

This guide shows you how to migrate your Teleport
cluster to use Proxy Peering, which enables you to scale
your Proxy Service instances horizontally by reducing the
number of connections created between Teleport Proxy instances
and Teleport services like the Database Service and Application Service.

## Prerequisites

An existing Teleport Enterprise cluster. See our [getting started guide](../../deploy-a-cluster/teleport-enterprise/getting-started.mdx) for help
setting up a Teleport cluster.

All components in the cluster should be running Teleport `10.0` or later. See our
[upgrade procedure](./upgrading.mdx) for help upgrading a Teleport cluster.

> Teleport Proxy instances must be able to reach each other over the network on port

Reviewer: ```suggestion
Teleport Proxy Service instances must be able to reach each other over the network on port
```

> 3021 by default. Ensure there are no firewall policies that would block communication

Reviewer: ```suggestion
`3021` by default. Ensure there are no firewall policies that would block communication
```

between instances.

## Step 1/3. Enable Proxy Peering
Update your cluster's Auth Service configuration to set the tunnel strategy type
to `proxy_peering`.

```yaml
auth_service:
    tunnel_strategy:
        type: proxy_peering
        agent_connection_count: 1
```

This setting will indicate to agents that they are only required to connect to 1
Teleport Proxy instance as specified by the `agent_connection_count` field.

For high availability, an `agent_connection_count` greater than 1 can be configured.

> This ensures an agent is still reachable if one of the proxies it is connected to is not available.

Reviewer: ```suggestion
This ensures an agent is still reachable if one of the Proxy Service instances it is connected to is not available.
```
